CP-1186 Automatic retrying for failed requests

evanweible-wf opened the pull request:
_Resolves #95_

## Issue
- #95 We should support configurable and transparent request retrying for failed requests

## Changes
**Source:**

### `AutoRetry`
- Add an `AutoRetry` class that encapsulates all configuration and information for a single request
  - Auto retry is opt-in (`enabled` flag must be set to `true`)
  - Max number of retries is configurable
  - Determining whether or not a request is retried is based on 3 factors (each of which has defaults and is configurable) that must each pass for the request to be eligible:
    - HTTP method
    - Response status code
    - Custom test (function that receives the finalized request and the response and returns true/false)
  - Contextual information is available (may be useful for interceptors to determine where in the lifecycle a request is, or for analytics, or for application code after a request finishes, etc)
    - # of attempts
    - list of failures
  - Auto retry is also not supported for certain requests
    - `StreamedRequest`s - can't read the stream body more than once
    - `MultipartRequest`s that contain files in the body - files' byte streams can't be read more than once

### `BaseRequest`
- Add a `clone()` method - necessary for requests
- `BaseRequest` provides a base implementation that constructs the correct request instance (taking into account whether or not it was created from a `Client`)
- Each implementing class of `BaseRequest` overrides this method, calls super to get the base clone, and then adds any request-specific details

### `CommonRequest`
- Support request retrying by testing a failed request for retry eligibility and attempting to retry a clone of the request. If the retry succeeds, that response is returned instead. If the retry eventually fails, that exception is returned.
- Switched to using a `Completer` for some of the async logic so that parts of the stack trace are not lost

### TODO
- [x] Add the `clone()` method to the browser implementation of `MultipartRequest` (currently missing)
- [x] Add a subset of the fields from the `AutoRetry` class to the `Client` class so that retry configuration can be easily applied to all requests
- [x] ~~Add some sort of logging/information when a request that would otherwise be eligible for retry is not retried because it is unsupported (see last bullet of `AutoRetry` above)~~ Deferring until https://github.com/Workiva/w_transport/issues/100
- [x] Update changelog

**Tests:**
- Unit tests added for several different variations of failed/successful requests with different auto-retry configurations.
- TODO:
  - [x] Test that request cancellation during a retry attempt functions correctly
  - [x] Add integration tests for request retrying
  - [x] Test `clone()` (base implementation and request-specific implementations)
  - [x] Test for attempting to retry streamed requests
  - [x] Test for attempting to retry multipart requests with files
  - [x] Test for requests created from a client and then cloned
  - [x] 100% coverage of added lines

## Areas of Regression
- HTTP request dispatch (should be adequately covered by tests)

## Testing
- CI passes
- Examples function as expected
- (Side note: I've created an issue to add an example that demonstrates request retrying https://github.com/Workiva/w_transport/issues/98)

## Code Review
@trentgrover-wf
@maxwellpeterson-wf 
@dustinlessard-wf
@jayudey-wf

rmconsole2-wf commented:
@evanweible-wf This pull request has merge conflicts, please resolve.

trentgrover-wf commented on lib/src/http/auto_retry.dart:
```diff
@@ -0,0 +1,105 @@
+import 'package:w_transport/src/http/response.dart';
+
+typedef /* Future<bool> | bool */ RetryTest(
+    FinalizedRequest request, BaseResponse response);
```
no need for this distinction here.  just assume the return type is `Future<bool>`.  if you do that, the async keyword is all that's necessary for a simple function to match the signature:
```
Future<bool> tryIt() async {
  return false;
}
```

evanweible-wf replied:
Yeah, I agree :/ Almost all use cases I can think of could be synchronous (async would only be necessary if a streamed response needed to be read), but it is simpler to just require async.

trentgrover-wf commented on lib/src/http/auto_retry.dart:
```diff
@@ -0,0 +1,105 @@
+  /// An additional test for a failed request's retry eligibility. This function
+  /// will receive the [FinalizedRequest] instance and the [BaseResponse]
+  /// instance from which a decision can be made. If the request should not be
+  /// retried, this should return `false` (or a `Future` that resolves to
```
let's just make the requirement = Future<bool>

evanweible-wf replied:
will do

trentgrover-wf commented on lib/src/http/auto_retry.dart:
```diff
@@ -0,0 +1,105 @@
+/// - [enabled] is true
+/// - [forHttpMethods] includes the HTTP method that the failed request used
+/// - [forStatusCodes] includes the response's status code
+/// - [test] (if not null) returns true
```
we should chat through the logic of this set of retry configurations.  it may impose too much configuration for simple use cases.  if all conditions must be met, does that mean you have to ensure that at least 3 of these things are properly configured?

evanweible-wf replied:
I'm still noodling on this

maxwellpeterson-wf commented on lib/src/http/common/client.dart:
```diff
@@ -93,7 +104,11 @@ abstract class CommonClient implements Client {
+      ..enabled = autoRetry.enabled
+      ..forHttpMethods = autoRetry.forHttpMethods
+      ..forStatusCodes = autoRetry.forStatusCodes
+      ..test = autoRetry.test;
```
Would it make sense to put this logic into an `AutoRetryConfig.clone()` method? Then you could do 

```dart
request.autoRetry = autoRetry.clone;
```

And avoid doing this in more than one place. Would make adding fields easier too, if that ever were to happen.

Just something to consider, this is fine too.

evanweible-wf replied:
I actually don't want to clone the auto retry object between retries - I want the same instance to be shared so that the contextual information (like the current number of attempts) is correct regardless of which request instance you have a reference to.

maxwellpeterson-wf commented:
One note to consider, but either way nice. +1

dustinlessard-wf commented:
Is there a way to retry with a specific duration between reattempts? Like, wait 500ms between reattempts?

evanweible-wf commented:
@dustinlessard-wf I hadn't thought of that, but I like that idea. I'll think about how best to add that (and I may defer it to a follow-up PR).

dustinlessard-wf commented:
:+1:

evanweible-wf commented:
@dustinlessard-wf created an issue to implement that later: https://github.com/Workiva/w_transport/issues/102

evanweible-wf commented:
@trentgrover-wf @maxwellpeterson-wf @dustinlessard-wf I reworked the retry eligibility test logic after some discussion around that area. Also updated the docs for the `AutoRetryConfig` class and added to the readme & changelog.

maxwellpeterson-wf commented:
+1 - nice

dustinlessard-wf commented:
+1

trentgrover-wf commented on lib/src/http/auto_retry.dart:
```diff
@@ -0,0 +1,147 @@
+/// The decision to retry a request or not can usually be made based solely on
+/// the request method and the response status code. For this reason, the
+/// [forHttpMethods] and [forStatusCodes] fields define a list of methods and
+/// status codes, respectively, for which it retrying is acceptable.
```
`for which retrying...` (get rid of extra `it`)

trentgrover-wf commented:
+1 pending minor doc fix

evanweible-wf commented:
@trentgrover-wf fixed doc string. Once you +1 that commit, I'll rebase and squash.

trentgrover-wf commented:
+1

dustinlessard-wf commented:
+1

evanweible-wf commented:
@trentgrover-wf @maxwellpeterson-wf @dustinlessard-wf rebased and squashed commits.

trentgrover-wf commented:
+1

maxwellpeterson-wf commented:
+1

evanweible-wf commented:
@jayudey-wf ready for merge

jayudey-wf commented:
QA Resource Approval: +10

- [ ] Testing instruction
- [ ] Dev +1's
- [ ] Dev/QA +10 with detail of what was tested
 - verified that examples were still working as expected
- [ ] Unit test created/updated
- [ ] All unit tests pass

Merging into master.
